package com.example.demo.domain;

import com.example.demo.validators.ValidDeletePart;
import org.hibernate.annotations.Columns;

import javax.persistence.*;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import java.io.Serializable;
import java.util.HashSet;
import java.util.Set;

/**
 *
 *
 *
 *
 */
@Entity
@ValidDeletePart
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name="part_type",discriminatorType = DiscriminatorType.INTEGER)
@Table(name="Parts")
public abstract class Part implements Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    long id;
    String name;
    @Min(value = 0, message = "Price value must be positive")
    double price;
    @Min(value = 0, message = "Inventory value must be positive")
    int inv;
    @Column(name = "min_inv", nullable = false)
    @Min(value = 0, message = "Inventory value must be zero or positive")
    Integer minInv;
    @Column(name = "max_inv", nullable = false)
    @Max(value = 500, message = "Max value must be between min and max values")
    Integer maxInv;

    @ManyToMany
    @JoinTable(name="product_part", joinColumns = @JoinColumn(name="part_id"),
            inverseJoinColumns=@JoinColumn(name="product_id"))
    Set<Product> products= new HashSet<>();

    public Part() {

    }

    public Part(String name, double price, int inv) {
        this.name = name;
        this.price = price;
        this.inv = inv;
    }

    public Part(long id, String name, double price, int inv) {
        this.id = id;
        this.name = name;
        this.price = price;
        this.inv = inv;
    }

    public Part(long id, String name, double price, int inv, Integer minInv, Integer maxInv) {
        this.id = id;
        this.name = name;
        this.price = price;
        this.inv = inv;
        this.minInv = minInv;
        this.maxInv = maxInv;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public double getPrice() {
        return price;
    }

    public void setPrice(double price) {
        this.price = price;
    }

    public int getInv() {
        return inv;
    }

    public void setInv(int inv) {
        this.inv = inv;
    }

    public void setMinInv(Integer minInv) {
        this.minInv = minInv;
    }

    public Integer getMinInv() {
        return minInv;
    }

    public void setMaxInv(Integer maxInv) {
        this.maxInv = maxInv;
    }

    public Integer getMaxInv() {
        return maxInv;
    }

    public boolean isValid() {

        if (minInv != null && maxInv != null && minInv > maxInv) {
            throw new RuntimeException("Invalid inventory values: Minimum Inventory cannot be greater than Maximum Inventory!");
        }
        if (minInv == null || maxInv == null) {
            throw new RuntimeException("Invalid inventory values: Minimum and Maximum Inventory must be set!");
        }
        if (inv < minInv) {
            throw new RuntimeException("Invalid inventory values: Inventory cannot be less than minimum Inventory!");
        }
        if (inv > maxInv) {
            throw new RuntimeException("Invalid inventory values: Inventory cannot be greater than maximum Inventory!");
        }
        return true;
    }

    public Set<Product> getProducts() {
        return products;
    }

    public void setProducts(Set<Product> products) {
        this.products = products;
    }

    public String toString(){
        return this.name;
    }
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Part part = (Part) o;

        return id == part.id;
    }

    @Override
    public int hashCode() {
        return (int) (id ^ (id >>> 32));
    }
}
